CIPs/0016.md

```diff
@@ -0,0 +1,89 @@
+# CIP [0016]: Standard celo protocol URI scheme
+
+- Date: 2020-07-27
+- Author: @yorhodes
+- Status: WIP
+
+## Overview
+
+Create a standard URI scheme for encoding interacting with the Celo protocol for various consumers including SDKs, mobile wallets, and end users.
+Facilitate development of QR systems by standardizing usage and unifying scheme for interoperability.
+
+## Goals
+
+- establish URI standard compatible with mobile app intents that optimizes for **memory** and **performance**
+- enable simple construction of URIs by humans without deep celo protocol understanding
+- enable simple construction of QRs by code from arbitrary celo transactions
+
+## Proposed Solution
+
+### Syntax
+Celo protocol URLs contain "celo" in their schema (protocol) part and are constructed as follows:
+
+    request                 = "celo:" [ prefix "-" ] target [ "/" endpoint ] [ "?" parameters ]
+    prefix                  = "pay" / "request" / "call" / "send"
+    
+    target                  = E164_PHONE_NUMBER / registry_id / celo_address
+    registry_id             = "StableToken" / "GoldToken" / "LockedGold" / ...
+    celo_address            = ( "0x" 40*40HEXDIG )
+    
+    endpoint                = token / function
+    token                   = "cUSD" / "CELO" / ...
+    function                = function_name "(" [ function_args ] ")"
+    function_name           = STRING
+    function_args           = function_arg *( "," function_arg )
+    function_arg            = ( NUMBER / STRING ) [ ":" TYPE ]
+
+    parameters              = parameter *( "&" parameter )
+    parameter               = key "=" value
+    key                     = "value" / "gas" / "gasPrice" / "feeCurrency" / "gatewayFee" / "gatewayFeeRecipient" / "localAmount" / "localCurrencyCode" / ...
```

Comment:
It would be nice to have an exhaustive list of keys that are valid and reject all keys that do not match